pageserver/src/virtual_file.rs

```diff
@@ -208,6 +208,78 @@ impl CrashsafeOverwriteError {
     }
 }
 
+/// Helper macro internal to `VirtualFile` that looks up the underlying File,
+/// opens it and evicts some other File if necessary. The passed parameter is
+/// assumed to be a function available for the physical `File`.
+///
+/// We are doing it via a macro as Rust doesn't support async closures that
+/// take on parameters with lifetimes.
```

Comment:
I think async closures are not supported at all, but `FnOnce(&VirtualFile) -> impl std::future::Future<Output = Result<T, E>>` seems like it could work.

Reply:
Aah oki, this wasn't about the body, which is invoked on a std::fs::File still. Where does the error come?

Reply:
Maybe it's the Seek, which I didn't remember was also implemented for `&File`, weirdly enough.

Seems this is implementable as, so I am unsure why the macro:

```rust
pub(crate) async fn with_file<F, T>(&self, op: F) -> Result<T, std::io::Error>
  where F: FnOnce(&mut &std::fs::File) -> Result<T, std::io::Error> {}
```

Reply:
Posted #5250 which seems to compile for me locally.

Reply:
Aah ok, so maybe this is all for future future use?

Reply:
Assuming this is just for future use, it still seems quite limited if you can only call methods on a file. Current plans have calling static fns with the file descriptor.

How about:

```
F: FnOnce(File) -> Fut,
Fut: std::future::Future<Output = (File, Result<T, std::io::Error>)>
```

Above would then work with:
- blocking calls to the file
- current async attempts
- maybe even tokio::fs?

Reply:
```rust
    pub(crate) async fn with_file<F, T>(&self, op: F) -> Result<T, std::io::Error>
    where
        F: FnOnce(&mut &std::fs::File) -> Result<T, std::io::Error>,
    {
```

#5250 is more or less the status quo: the functions we want to call will one day be async fn:


> Aah ok, so maybe this is all for future future use?

Yeah, that's what this PR is about (and all of my prior PRs for #4743 :) ). Quoting the issue's DOD:

> All code up to the point where we issue the system call to read from Image or Delta layer files is `async`.
The system call itself remains synchronous, it is _not_ converted to `tokio::fs`.
This allows merging the work in this PR without performance regressions caused by `tokio::fs`, thereby isolating any potential performance regressions to any refactoring changes required.

>  it still seems quite limited if you can only call methods on a file. Current plans have calling static fns with the file descriptor.

I don't know how else to pass the file to the snippet than via dot syntax. hygiene doesn't allow you to just access the `file` variable. I haven't tried out closures, as weird as it sounds they might work. It *might* be possible to change it to create a list of args or something, but what is possible in any case is to add a thin wrapper over `File` with a bunch of inherent functions.

Reply:
Perhaps Heikki's proposal here is a better version? https://github.com/neondatabase/neon/pull/5247#issuecomment-1711200985

Reply:
Yeah guards might work and I think they would be nicer than a macro (if they work).

Reply:
> Where does the error come?

It's the same issue as in #5224 for the `test_files` function. See my comment there: https://github.com/neondatabase/neon/pull/5224#issuecomment-1710255989 . The error is quite analogous here:

```
error: lifetime may not live long enough
   --> pageserver/src/virtual_file.rs:516:39
    |
516 |             .with_file("read", |file| async move { file.read_at(buf, offset) })
    |                                 ----- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                 |   |
    |                                 |   return type of closure `[async block@pageserver/src/virtual_file.rs:516:39: 516:79]` contains a lifetime `'2`
    |                                 has type `&'1 std::fs::File`
```

If you have:

```Rust
    async fn with_file<F, R, FR>(&self, _op: &str, func: F) -> Result<R, Error>
    where
        F: FnOnce(&File) -> FR,
        FR: Future<Output = R>,
    {
```

See my branch [virtual_file_async_5_5](https://github.com/neondatabase/neon/compare/arpad/virtual_file_async_5_5) if you want to try it out yourself.

Reply:
>  I haven't tried out closures, as weird as it sounds they might work. 

They don't: :(

```
   --> pageserver/src/virtual_file.rs:520:53
    |
520 |         let result = with_file!(self, |file: &File| async { file.write_at(buf, offset) });
    |                                              -    - ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                              |    |
    |                                              |    return type of closure `[async block@pageserver/src/virtual_file.rs:520:53: 520:89]` contains a lifetime `'2`
    |                                              let's call the lifetime of this reference `'1`

``

Reply:
I think this is good to close now because guard is used, though, together with the macro, but agreed, it does not work with `|file: &File|` we'd have to have `Arc<File>`.

Reply:
Well, this is a better thread to continue than to start a new.

I still dislike the tt-matching macro for this simply because we cannot use for example tokio::fs::File or something else, so this will need to be resolved very soon after this PR..

Which I guess could be fine, but feels like we could find a way out on this PR.

Could we tt-munch a fake async closure? match `|$identifier:ident| async { ... }` then match that as `|$identifier| $body` then execute that as 

```rust
let res = {
  let $identifier = sl.as_ref();
  async { $body }.await;
};
```

(I don't know how to implement this, but I am assuming you do :)) It might actually be that the hygiene rules prohibit that.. If it were to work then we'd be able to express tokio::fs and and another way for this.

Though, I cannot see a way out the "other ways" from requiring the fd to be `Arc<OwnedFd|File>` because it needs to stay valid for the duration of the async operation which could be done on a userspace or kernel thread pool.

Reply:
I will try out your idea, it sounds good :+1: